data-contracts/README.md

```diff
@@ -0,0 +1,439 @@
+terraform apply "tfplan"
+```
+
+Let's have a look at what we've created. We now have a Confluent Cloud environment, which includes a Kafka cluster and Data Governance platform with Confluent Schema Registry. Within the Kafka cluster, we create a Kafka topic named `membership-avro`.
+
+```terraform
+resource "confluent_kafka_topic" "membership_avro" {
+  topic_name = "membership-avro"
+  kafka_cluster {
+    id = confluent_kafka_cluster.kafka_cluster.id
+  }
+  rest_endpoint = confluent_kafka_cluster.kafka_cluster.rest_endpoint
+...
+  partitions_count = 10
+...
+}
+```
+
+Then we configure the subject containing the value of events for this topic using `confluent_subject_config`. This tells the Schema Registry that `membership-avro-value` schemas will be `BACKWARD` compatible and defines an attribute called `major_version` to be used as the `compatibilityGroup`.
+
+```terraform
+resource "confluent_subject_config" "membership_value_avro" {
+  subject_name = "${confluent_kafka_topic.membership_avro.topic_name}-value"
+
+  schema_registry_cluster {
+    id = data.confluent_schema_registry_cluster.advanced.id
+  }
+  rest_endpoint = data.confluent_schema_registry_cluster.advanced.rest_endpoint
+...
+  compatibility_level = "BACKWARD"
+  compatibility_group = "major_version"
+...  
+}
+```
+
+For more on these data contract configuration enhancements, refer to the CC docs on [data contracts](https://docs.confluent.io/cloud/current/sr/fundamentals/data-contracts.html#configuration-enhancements). 
+
+
+#### Prepare Client Properties
+
+This demo has tailored the output of `terraform apply` to return the properties needed to connect to Confluent Cloud. The command below will 
+reformat the names of those properties into the names used in Kafka Client configurations, then export those outputs to a properties file 
+in our project:
+
+```shell
+terraform output -json | \
+  jq -r 'to_entries | map( {key: .key|tostring|split("_")|join("."), value: .value} ) | map("\(.key)=\(.value.value)")' | while read -r line ; do echo "$line"; \
+  done > ../shared/src/main/resources/confluent.properties
+```
+
+All Kafka Client code in this project loads connection properties form `shared/src/main/resources/confluent.properties`. For an example of this 
+properties file, see [confluent.properties.orig](shared/src/main/resources/confluent.properties.orig).
+
+> [!NOTE]
+> The file-based approach we're using here is NOT recommended for a production-quality application. Perhaps a secrets manager implementation would be better suited - which the major cloud providers all offer, or perhaps a tool like Hashicorp Vault. Such a tool would also have client libraries in a Maven repository for the JVM applications to access the secrets.
+
+### Producing and Consuming Events
+
+We'll create producer and consumer classes to configure and provide Kafka clients for our application(s). In the `shared` module there are implementations to encapsulate this behavior:
+
+```kotlin
```

Comment:
thinking of ways to trim down the README and think we don't need the code snippet. say that the clients are dummy implementations, e.g., the client simply logs events, but dont need to include the code in a README

Reply:
Would making these "collapsible" work a lil better?

Reply:
my pref would be a link (non-relative for out friend CD) to the source for this particular one since the producer / consumer are kind of background noise far from the point of this tutorial. my high level 2c on snippets by example - definitely do show a snippet and explain the client configs that are relevant to the tutorial topic (`LATEST_COMPATIBILITY_STRICT`, `USE_LATEST_VERION`, `USE_LATEST_WITH_METADATA`), but inlining code snippets of the class hierarchy and showing a consumer that logs is too much detail and not relevant enough to make the README cut IMO... the snippets kinda detract from the thing to highlight which is those three configs